### PR TITLE
*: abstract the Importer communication into an interface

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Turn the `*kv.Importer` struct into an abstract interface to prepare for the MySQL (Loader) backend.

### What is changed and how it works?

Introduced the "backend" abstraction at `backend.go`.

1. Introduced an `AbstractBackend` interface to replace the `ImportKVClient` gRPC interface. Some APIs contain extra arguments that would be used by the MySQL backend.
2. The `AbstractBackend` interface is wrapped into a `Backend` struct to provide the concrete APIs which performs logging and retry.
3. Since the SQL → KV step is not needed in the MySQL backend, the Encoder interface is also abstracted. The output of the encoder is a Row interface, and the collection is a Rows interface.

| Interface / Method | Importer | MySQL (Provisional) |
|---|---|---|
| (AbstractBackend).RetryImportDelay() | 3s | 0ns (no import step) |
| (AbstractBackend).MaxChunkSize() | 31 MiB | (to be decided) |
| (AbstractBackend).OpenEngine() | OpenEngine | no-op |
| (AbstractBackend).WriteRows() | WriteEngine | INSERT INTO ... |
| (AbstractBackend).CloseEngine() | CloseEngine | no-op |
| (AbstractBackend).ImportEngine() | ImportEngine | no-op |
| (AbstractBackend).CleanupEngine() | CleanupEngine | no-op |
| Encoder | TableKVEncoder | no-op encoder |
| Row | `[]kvenc.KvPair` | `[]types.Datum` |
| Rows | `[]kvenc.KvPair` | `[][]types.Datum` |

Importer supports `SwitchMode` and `Compact` by directly forwarding the command to the TiKV stores. We cut the middleman by issuing the commands directly from Lightning, thus reducing the API surface of Backend. These are placed into `kv/tikv.go`.

Some tests in the `restore` package are moved to the `kv` package as some algorithms are placed into the Backend/Encoder/Row/Rows interface directly. 

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

### CODE REVIEW

**Comment by GregoryIan** [comment]
@leoppro PTAL

**Comment by GregoryIan** [comment]
was CI always blocked?

**Comment by kennytm** [comment]
@GregoryIan  Currently the CI needs to be executed manually 😅 

/run-all-tests

File: lightning/restore/restore.go [hunk 1]
```diff
@@ -933,7 +931,15 @@ func (rc *RestoreController) fullCompact(ctx context.Context) error {
 }
 
 func (rc *RestoreController) doCompact(ctx context.Context, level int32) error {
-	return errors.Trace(rc.importer.Compact(ctx, level))
+	return kv.ForAllStores(
+		ctx,
+		&http.Client{},
+		rc.cfg.TiDB.PdAddr,
+		kv.StoreStateUp,
```
Comment (GregoryIan) [inline lightning/restore/restore.go]: is it ok?

File: lightning/kv/importer.go [hunk 2]
```diff
@@ -166,98 +97,92 @@ func isIgnorableOpenCloseEngineError(err error) bool {
...
+		return nil
+	}
+
+	wstream, err := importer.cli.WriteEngine(ctx)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return errors.Trace(err)
 	}
 
+	logger := log.With(zap.Stringer("engineUUID", engineUUID))
```
[the remark this answers is not in the document]
Comment (kennytm) [reply lightning/kv/importer.go]: @GregoryIan what does the field log?

[the remark this answers is not in the document]
Comment (GregoryIan) [reply lightning/kv/importer.go]: like `logger := log.With(zap.Stringer("engineUUID", engineUUID), zap.String("step", "stream write"))`, but it‘s not a big problem

File: cmd/tidb-lightning-ctl/main.go [hunk 3]
```diff
@@ -102,17 +103,15 @@ func run() error {
...
-	if err := importer.Compact(ctx, restore.FullLevelCompact); err != nil {
-		return errors.Trace(err)
-	}
-
-	return nil
+	return kv.ForAllStores(
+		ctx,
+		&http.Client{},
+		cfg.TiDB.PdAddr,
+		kv.StoreStateOffline,
```
Comment (GregoryIan) [inline cmd/tidb-lightning-ctl/main.go]:
```	
// StoreStateDisconnected means the TiKV store does not respond to PD.
StoreStateDisconnected
```

what if some tikvs would be in short-lived disconnection, and come back soon？

**Comment by GregoryIan** [comment]
Rest LGTM

File: lightning/kv/importer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kennytm) [reply lightning/kv/importer.go]: Not gonna change this since the log message should already be clear enough which step we're in.

**Pull request merged**